### PR TITLE
Consolidate Graphics properties, fixing compat bugs

### PR DESCRIPTION
**Pull request by hughbe**

### CODE REVIEW

File: src/System.Drawing.Common/src/System/Drawing/Font.Unix.cs [hunk 1]
```diff
@@ -224,27 +84,18 @@ public static Font FromHfont(IntPtr hfont)
...
-
-            return (new Font(newObject, lf.lfFaceName, newStyle, newSize));
+            return new Font(newObject, lf.lfFaceName, newStyle, Math.Abs(lf.lfHeight));
         }
 
         public IntPtr ToHfont()
         {
-            if (fontObject == IntPtr.Zero)
+            if (_nativeFont == IntPtr.Zero)
                 throw new ArgumentException("Object has been disposed.");
```
Comment (safern) [inline src/System.Drawing.Common/src/System/Drawing/Font.Unix.cs]: Now that you're in this cleanup should we change `"Object has been disposed."` to use a resource string?

File: src/System.Drawing.Common/tests/FontTests.cs [hunk 2]
```diff
@@ -725,6 +711,12 @@ public void FromLogFont_InvalidLogFont_ThrowsArgumentException()
         [InlineData(GraphicsUnit.World)]
         public void SizeInPoints_Get_ReturnsExpected(GraphicsUnit unit)
         {
+            if (PlatformDetection.IsRedHatFamily)
+            {
+                // Returns NaN on RedHat.
```
Comment (safern) [inline src/System.Drawing.Common/tests/FontTests.cs]: Ditto.

**Comment by hughbe** [comment]
Rebased onto https://github.com/dotnet/corefx/pull/29673

**Comment by safern** [comment]
@dotnet-bot test this please

File: src/System.Drawing.Common/src/System/Drawing/Font.Unix.cs [hunk 3]
```diff
@@ -224,27 +84,20 @@ public static Font FromHfont(IntPtr hfont)
...
+            {
+                throw new ArgumentException(SR.ObjectDisposed);
+            }
 
-            return fontObject;
+            return _nativeFont;
         }
 
-        internal Font(IntPtr newFontObject, string familyName, FontStyle style, float size)
+        internal Font(IntPtr new_nativeFont, string familyName, FontStyle style, float size)
```
Comment (stephentoub) [inline src/System.Drawing.Common/src/System/Drawing/Font.Unix.cs]: Nit: this is a strange name for an argument... we usually don't use underscores like this.

Comment (hughbe) [reply src/System.Drawing.Common/src/System/Drawing/Font.Unix.cs]: Fixed in https://github.com/dotnet/corefx/pull/29673

**Comment by hughbe** [comment]
No longer relies on https://github.com/dotnet/corefx/pull/29673/files (so can be merged seperately and ideally before)

File: src/System.Drawing.Common/src/System/Drawing/Graphics.cs [hunk 4]
```diff
@@ -317,6 +609,10 @@ public void MultiplyTransform(Matrix matrix, MatrixOrder order)
             {
                 throw new ArgumentNullException(nameof(matrix));
             }
+            if (matrix.nativeMatrix == IntPtr.Zero)
+            {
+                return;
+            }
```
Comment (stephentoub) [inline src/System.Drawing.Common/src/System/Drawing/Graphics.cs]: Why was this added?

File: src/System.Drawing.Common/src/System/Drawing/Graphics.cs [hunk 5]
```diff
@@ -20,6 +21,369 @@ public sealed partial class Graphics : MarshalByRefObject, IDisposable, IDeviceC
...
+                        value = SmoothingMode.None;
+                        break;
+                    case SmoothingMode.HighQuality:
+                        value = SmoothingMode.AntiAlias;
+                        break;
+                    case SmoothingMode.Invalid:
+                        throw new ArgumentException(SR.GdiplusInvalidParameter);
+                    default:
+                        break;
+                }
```
Comment (stephentoub) [inline src/System.Drawing.Common/src/System/Drawing/Graphics.cs]: Where did this switch block come from, and why?  It's not in the original Windows code; why is it needed here?

File: src/System.Drawing.Common/src/System/Drawing/Graphics.cs [hunk 6]
```diff
@@ -20,6 +21,369 @@ public sealed partial class Graphics : MarshalByRefObject, IDisposable, IDeviceC
...
+            {
+                if (value < PixelOffsetMode.Invalid || value > PixelOffsetMode.Half)
+                {
+                    throw new InvalidEnumArgumentException(nameof(value), unchecked((int)value), typeof(PixelOffsetMode));
+                }
+
+                if (value == PixelOffsetMode.Invalid)
+                {
+                    throw new ArgumentException(SR.GdiplusInvalidParameter);
+                }
```
Comment (stephentoub) [inline src/System.Drawing.Common/src/System/Drawing/Graphics.cs]: Where did this extra if block come from, and why?

File: src/System.Drawing.Common/src/System/Drawing/Graphics.cs [hunk 7]
```diff
@@ -20,6 +21,369 @@ public sealed partial class Graphics : MarshalByRefObject, IDisposable, IDeviceC
...
+                        value = InterpolationMode.Bilinear;
+                        break;
+                    case InterpolationMode.High:
+                        value = InterpolationMode.HighQualityBicubic;
+                        break;
+                    case InterpolationMode.Invalid:
+                        throw new ArgumentException(SR.GdiplusInvalidParameter);
+                    default:
+                        break;
+                }
```
Comment (stephentoub) [inline src/System.Drawing.Common/src/System/Drawing/Graphics.cs]: I don't understand why all of this new code is being added.  It's a little frustrating to review, as new code appears to be coming out of nowhere in a "consolidation" PR, and is changing well-established code paths on Windows.

File: src/System.Drawing.Common/src/System/Drawing/Graphics.cs [hunk 8]
```diff
@@ -20,6 +21,369 @@ public sealed partial class Graphics : MarshalByRefObject, IDisposable, IDeviceC
...
+            {
+                if (value < GraphicsUnit.World || value > GraphicsUnit.Millimeter)
+                {
+                    throw new InvalidEnumArgumentException(nameof(value), unchecked((int)value), typeof(GraphicsUnit));
+                }
+
+                if (value == GraphicsUnit.World)
+                {
+                    throw new ArgumentException(SR.GdiplusInvalidParameter);
+                }
```
Comment (stephentoub) [inline src/System.Drawing.Common/src/System/Drawing/Graphics.cs]: Ditto

File: src/System.Drawing.Common/src/System/Drawing/Graphics.cs [hunk 9]
```diff
@@ -20,6 +21,369 @@ public sealed partial class Graphics : MarshalByRefObject, IDisposable, IDeviceC
...
+
+                return scale[0];
+            }
+
+            set
+            {
+                if (value <= 0 || value > 1000000032)
+                {
+                    throw new ArgumentException(SR.GdiplusInvalidParameter);
+                }
```
Comment (stephentoub) [inline src/System.Drawing.Common/src/System/Drawing/Graphics.cs]: Ditto

Comment (hughbe) [reply src/System.Drawing.Common/src/System/Drawing/Graphics.cs]: This fixes the test `PageScale_SetInvalid_ThrowsArgumentException` with libgdiplus

File: src/System.Drawing.Common/src/System/Drawing/Graphics.cs (hunk 8 above)
Comment (hughbe) [reply src/System.Drawing.Common/src/System/Drawing/Graphics.cs]: This fixes the test `PageUnit_SetWorld_ThrowsArgumentException` with libgdiplus

File: src/System.Drawing.Common/src/System/Drawing/Graphics.cs (hunk 7 above)
Comment (hughbe) [reply src/System.Drawing.Common/src/System/Drawing/Graphics.cs]: This fixes the test `InterpolationMode_SetValid_GetReturnsExpected` with libgdiplus

File: src/System.Drawing.Common/src/System/Drawing/Graphics.cs (hunk 6 above)
Comment (hughbe) [reply src/System.Drawing.Common/src/System/Drawing/Graphics.cs]: This fixes the test `PixelOffsetMode_SetToInvalid_ThrowsArgumentException` with libgdiplus

File: src/System.Drawing.Common/src/System/Drawing/Graphics.cs (hunk 5 above)
Comment (hughbe) [reply src/System.Drawing.Common/src/System/Drawing/Graphics.cs]: This fixes the test `SmoothingMode_SetValid_GetReturnsExpected` with libgdiplus

File: src/System.Drawing.Common/src/System/Drawing/Graphics.cs (hunk 4 above)
Comment (hughbe) [reply src/System.Drawing.Common/src/System/Drawing/Graphics.cs]: This fixes the test `MultiplyTransform_DisposedMatrix_Nop` with libgdiplus

**Comment by hughbe** [comment]
Sorry that I mixed in compat fixes with the consolidation commit - I should've been more clear

File: src/System.Drawing.Common/src/System/Drawing/Graphics.cs (hunk 7 above)
Comment (stephentoub) [reply src/System.Drawing.Common/src/System/Drawing/Graphics.cs]: For all of these, it sounds like the bug is in libgdiplus (i.e. it's exhibiting different behavior than the Windows library it's emulating), rather than something we should be working around by modifying a code path that also impacts Windows.  I'd rather see libgdiplus fixed and the relevant tests special-cased for Unix until we absorb that fix.

Comment (hughbe) [reply src/System.Drawing.Common/src/System/Drawing/Graphics.cs]: In most of these cases, libgdiplus behaviour has been fixed. However, I believe that there is an existing precedent for updating the csharp product code to reflect the fixes in libgdiplus knowing that many systems do not have the latest libgdiplus. @mellinoe and @safern may be able to comment. For example, in `LinearGradientBrush.cs` we have done the same:

> // Do explicit parameter validation here; libgdiplus does not correctly validate the arguments

```cs
            // Multiplying the transform by a disposed matrix is a nop in GDI+, but throws
            // with the libgdiplus backend. Simulate a nop for compatability with GDI+.

```

Comment (stephentoub) [reply src/System.Drawing.Common/src/System/Drawing/Graphics.cs]: I'll leave it up to @safern.  I don't like that we're "harming" the code that's doing what it's supposed to be doing when the platform's working as it should, but if we've already established a precedent for that elsewhere (and we did so purposefully) and @safern wants to continue going in that direction, ok.

Comment (safern) [reply src/System.Drawing.Common/src/System/Drawing/Graphics.cs]: I agree that it is hard to review consolidating changes and it indeed makes it harder to review whenever there is code fixes. We've followed this pattern in some other places where we do stuff in our side to reflect a fix in libgdiplus, so I don't mind doing that to make the library more stable in unix systems, if we say we support many distros I think we should try to support those platforms and be consistent and functional across them. From what I know in some places we do some complicated dances in Linux or BSD, depending on the platform to try to be functional across platforms in other libraries. 

What I don't like here is to mix up those fixes with the windows code. For example in this case, if GDI+ works fine, why do it in the shared code and not doing this in the Unix code?

Comment (safern) [reply src/System.Drawing.Common/src/System/Drawing/Graphics.cs]: And BTW thanks @stephentoub for reviewing such big PRs in System.Drawing lately :)

Comment (hughbe) [reply src/System.Drawing.Common/src/System/Drawing/Graphics.cs]: > What I don't like here is to mix up those fixes with the windows code. For example in this case, if GDI+ works fine, why do it in the shared code and not doing this in the Unix code?

Simply because this is what we've done already - almost all cases consolidating code between mono and Windows have started with the baseline of moving the code to the same file and adding argument validation to the shared code

**Comment by stephentoub** [comment]
If you're ok with it, it's fine.
